### PR TITLE
Team Bob

### PR DESCRIPTION
**Pull request by lizy14**

### CODE REVIEW

File: adminpage/views.py [hunk 1]
```diff
@@ -1,3 +1,272 @@
...
+            return SITE_DOMAIN + '/media/img/' + image.name
+        except:
+            raise ValidateError('Fail to upload image')
+
+
+class ActivityCheckin(APIView):
+    def post(self):
+        if not self.request.user.is_authenticated():
+            raise ValidateError("User not logged in!")
+        self.check_input('actId')
```
Comment (rjswjs123) [inline adminpage/views.py:246]: self.check_input('ticket')

File: adminpage/views.py [hunk 2]
```diff
@@ -1,3 +1,272 @@
...
+        try:
+            activity = Activity.objects.get(id=self.input['actId'])
+        except:
+            raise LogicError('actID error！Activity not found')
+        try:
+            if "ticket" in self.input:
+                ticket = Ticket.objects.get(activity=activity,unique_id=self.input['ticket'])
+            elif 'studentId' in self.input.keys():
+                ticket = Ticket.objects.get(activity=activity,student_id=self.input['studentId'])
```
Comment (rjswjs123) [inline adminpage/views.py:255]: ticket.objects.filter

File: wechat/handlers.py [hunk 3]
```diff
@@ -1,10 +1,20 @@
...
+from datetime import datetime
+import time
+import random
+import string
+
 
 
 __author__ = "Epsirom"
 
+remainTicketsLock = threading.Lock()
```
Comment (fang-zh16) [inline wechat/handlers.py:17]: nice code here!

File: adminpage/views.py [hunk 4]
```diff
@@ -1,3 +1,272 @@
...
+            end_time = datetime.strptime(self.input['endTime'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8),
+            book_start = datetime.strptime(self.input['bookStart'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8),
+            book_end = datetime.strptime(self.input['bookEnd'], "%Y-%m-%dT%H:%M:%S.%fZ") + timedelta(hours=8),
+            remain_tickets = self.input['totalTickets'],
+            total_tickets = self.input['totalTickets'],
+            status = self.input['status']
+        )
+        new_activity.save()
+        return new_activity.id
+
```
Comment (rjswjs123) [inline adminpage/views.py:108]: try:
 new_activity = Activity()
except:
raise ValidateError('Create activity error!')

**Pull request closed without merging**